custom_random_forest.py

```diff
@@ -0,0 +1,172 @@
+from sklearn.base import BaseEstimator
+from sklearn.tree import DecisionTreeClassifier
+from joblib import Parallel, delayed
+import numpy as np
+
+# HW18
+SEED = 42
+
+
+class RandomForestClassifierCustom(BaseEstimator):
+    """
+    Custom implementation of a Random Forest Classifier.
+
+    Parameters
+    ----------
+    n_estimators : int, default=10
+        The number of trees in the forest.
+
+    max_depth : int, default=None
+        The maximum depth of the tree. If None, then nodes are expanded until all leaves are pure or until all leaves contain less than min_samples_split samples.
+
+    max_features : int, default=None
+        The number of features to consider when looking for the best split.
+
+    random_state : int, default=SEED
+        Controls the randomness of the bootstrapping of the samples used when building trees.
+    """
+
+    def __init__(
+        self, n_estimators=10, max_depth=None, max_features=None, random_state=SEED
+    ):
+        self.n_estimators = n_estimators
+        self.max_depth = max_depth
+        self.max_features = max_features
+        self.random_state = random_state
+
+        self.trees = []
+        self.feat_ids_by_tree = []
+
+    def _train_single_tree(self, i, X, y):
+        """
+        Trains a single decision tree.
+
+        Parameters
+        ----------
+        i : int
+            Index of the tree being trained.
+
+        X : array-like of shape (n_samples, n_features)
+            The training input samples.
+
+        y : array-like of shape (n_samples,)
+            The target values.
+
+        Returns
+        -------
+        tuple
+            A tuple containing the feature indices used and the trained decision tree.
+        """
+        np.random.seed(self.random_state + i)
```

Comment:
Я бы только отметил что `i` это же ведь не просто немой счетчик
```suggestion
        np.random.seed(self.random_state + tree_id)
```